EIPS/eip-5380.md

```diff
@@ -0,0 +1,91 @@
+
+## Specification
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+### `IERC5380`
+
+```solidity
+/// SPDX-License-Identifier: CC0-1.0
+pragma solidity ^0.8.0;
+
+interface IERC5380 is IERC165 {
+    /// @notice Emitted when the amount of entitlement a user changes. If user is the zero address, then the user is the owner
+    event Entitle(address indexed user, address indexed contract, uint256 indexed tokenId);
+
+    /// @notice             Set the user associated with the given EIP-721 token as long as the owner is msg.sender.
+    /// @dev                SHOULD NOT revert if the owner is not msg.sender.
+    /// @param  user        The user to grant the entitlement to
+    /// @param  contract    The property to grant
+    /// @param  tokenId     The tokenId to grant the properties of
+    function entitle(address user, address contract, uint256 tokenId) external;
+
+    /// @notice             Get the maximum number of users that can receive this entitlement
+    /// @param  contract    The contract to query
+    /// @param  tokenId     The tokenId to query
+    function maxEntitlements(address contract, uint256 tokenId) external view (uint256 max);
+
+    /// @notice             Get the user associated with the given contract and tokenId.
+    /// @dev                Defaults to contract.ownerOf(tokenId)
+    /// @param  contract    The contract to query
+    /// @param  tokenId     The tokenId to query
+    function entitlementOf(address contract, uint256 tokenId) external view returns (address user);
+}
+```
+
+`supportsInterface` MUST return true when called with `IERC5380`'s interface ID.
+
+### `IERC5380Enumerable`
+
+This OPTIONAL interface is RECOMMENDED.
+
+```solidity
+/// SPDX-License-Identifier: CC0-1.0
+pragma solidity ^0.8.0;
+
+interface IERC5380Enumerable is IERC5380 /* , IERC165 */ {
+    /// @notice         Enumerate tokens assigned to a user
+    /// @dev            Throws if the index is out of bounds or if user == address(0)
+    /// @param  user    The user to query
+    /// @param  index   A counter
+    function tokenOfUserByIndex(address user, uint256 index) external view returns (address contract, uint256 tokenId);
+}
+```
+
+`supportsInterface` MUST return true when called with `IERC5380Enumerable`'s interface ID.
+
+## Rationale
+
+TODO.
+
+## Backwards Compatibility
+
+No backward compatibility issues were found.
```

Comment:
[Optional] If this EIP adds `uint256 amount` in the interfaces, it could have the benefit of supporting *both* EIP-721 and EIP-1155, potentially helpful for its adoption.

Reply:
The issue here is supporting `maxEntitlements`. I initially was trying to do it in a cross-token manner, but it just didn't particularly work well. It's not clear who loses what when the balance changes without going to zero.